python/ray/dag/compiled_dag_node.py

```diff
@@ -803,9 +821,14 @@ def _preprocess(self) -> None:
 
                 self.actor_task_count[actor_handle._actor_id] += 1
 
-                if dag_node.type_hint.requires_nccl():
+                if not isinstance(dag_node, CollectiveOutputNode):
                     # Add all writers to the NCCL group.
-                    nccl_actors.add(actor_handle)
+                    if dag_node.type_hint.requires_nccl():
+                        nccl_actors.add(actor_handle)
+                elif isinstance(dag_node, CollectiveOutputNode):
+                    # Initialize the NCCL group on the participating actors
+                    # for collective methods.
+                    dag_node._init_nccl_group()
```

Comment:
Here, we need to deduplicate the NCCL groups.

Reply:
If the NCCL group is a custom group ([GPUCommunicator](https://github.com/ray-project/ray/pull/47141/files#diff-747dfde558716ed96f5b860673ae997fcdb0258bf4735467bbe5576a67225737)), then we also need to check the membership of the group against the nodes that are included in the collective. If the memberships don't match, throw an error.

Reply:
Done.